### PR TITLE
Move Pane to Tab (GH7075)

### LINKED ISSUE #7075: Allow moving panes to other tabs
# Description of the new feature/enhancement

Simply put, I would like to be able to take a single pane and either transform it into a tab, or add it as a pane of a different tab.
In my opinion this is useful when a specific tab gets overcrowded with panes, and gets cumbersome.
You could close the panes you want to move and reopen new ones in a different tab, but then you lose output of previous commands, environment changes you made, etc.

As for how this might appear to the user, I think it would have to be a drag & drop kind of thing for the use case of moving the pane to an existing tab (Also allowing to rearrange panes in the same tab which would be nice). Keyboard shortcuts might work too, but I think it'll make it hard to reach the exact arrangement of tabs.

### PR DESCRIPTION
**Pull request by Rosefield**
## Summary of the Pull Request
Add functionality to move a pane to another tab. If the tab index is greater than the number of current tabs a new tab will be created with the pane as its root. Similarly, if the last pane on a tab is moved to another tab, the original tab will be closed.

This is largely complete, but I know that I'm messing around with things that I am unfamiliar with, and would like to avoid footguns where possible. 

 
## References
#4587 

## PR Checklist
* [x] Closes #7075
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [x] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [x] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments
Things done:
- Moving a pane to a new tab appears to work. Moving a pane to an existing tab mostly works. Moving a pane back to its original tab appears to work.
- Set up {Attach,Detach}Pane methods to add or remove a pane from a pane. Detach is slightly different than Close in that we want to persist the tree structure and terminal controls.
- Add `Detached` event on a pane that can be subscribed to to remove other event handlers if desired. 
- Added simple WalkTree abstraction for one-off recursion use cases that calls a provided function on each pane in order (and optionally terminates early).
- Fixed an in-prod bug with closing panes. Specifically, if you have a tree (1; 2 3) and close the 1 pane, then 3 will lose its borders because of these lines clearing the border on both children https://github.com/microsoft/terminal/blob/main/src/cascadia/TerminalApp/Pane.cpp#L1197-L1201 .

To do:
- Right now I have `TerminalTab` as a friend class of `Pane` so I can access some extra properties in my `WalkTree` callbacks, but there is probably a better choice for the abstraction boundary.

Next Steps:
- In a future PR Drag & Drop handlers could be added that utilize the Attach/Detach infrastructure to provide a better UI.
- Similarly once this is working, it should be possible to convert an entire tab into a pane on an existing tab (Tab::DetachRoot on original tab followed by Tab::AttachPane on the target tab).
- Its been 10 years, I just really want to use concepts already.

## Validation Steps Performed
Manual testing by creating pane(s), and moving them between tabs and creating new tabs and destroying tabs by moving the last remaining pane.

### CODE REVIEW

**Comment by zadjii-msft** [comment]
fyi the team probably won't get around to reviewing PRs this week. I don't want you to get the impression we're ignoring you 😝

**Comment by Rosefield** [comment]
merged in `main` last night which fixed all of the CI errors. This should be fully ready now.

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Okay, I think this looks great. I need to socialize the commandline here a bit with the rest of the team to see what they think. Overall, I think this is a great PR and sets us up well for some really great features in the future!

(only blocking for the team discussion, usually mondays)

File: src/cascadia/TerminalApp/AppCommandlineArgs.cpp [hunk 1]
```diff
@@ -297,6 +298,43 @@ void AppCommandlineArgs::_buildSplitPaneParser()
...
+// - Adds the `move-pane-to-tab` subcommand and related options to the commandline parser.
+// - Additionally adds the `mptt` subcommand, which is just a shortened version of `move-pane-to-tab`
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void AppCommandlineArgs::_buildMovePaneToTabParser()
+{
+    _movePaneToTabCommand = _app.add_subcommand("move-pane-to-tab", RS_A(L"CmdMovePaneToTabDesc"));
+    _movePaneToTabShort = _app.add_subcommand("mptt", RS_A(L"CmdMPTTDesc"));
```
[the remark this answers is not in the document]
Comment (Rosefield) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:311]: One of the other things I was thinking about while implementing this is, especially forward looking, that you could have a target pane, and a `SplitStyle`. E.g. if directional split and drag and drop is added, the underlying command might be like "move pane to tab 2, with target pane 5, split and place on the left side." That is at least a couple of features away, so I don't know how much it makes sense it build that in quite yet. 

As for the command line interface for such a feature, I'm not sure if it needs to be so involved. As a vim user, I like to think of commands and movements. things like `move-focus` and `move-pane` are commands that operate on a movement, whereas "move-pane-to-tab" seems more self contained. Maybe I'm just not creative enough and "move-pane-to-tab previous" or "move-pane-to-tab down" are a suitable interface.

File: doc/cascadia/profiles.schema.json [hunk 2]
```diff
@@ -493,6 +494,23 @@
               "type": "integer",
               "default": 0,
               "description": "Which tab to switch to, with the first being 0"
+          }
+        }
+      }
+    ],
+    "required": [ "index" ]
+  },
+  "MovePaneToTabAction": {
```
[the remark this answers is not in the document]
Comment (Rosefield) [reply doc/cascadia/profiles.schema.json:503]: I added it there, don't really know how to confirm correctness of anything in the schema though.

**Comment by Rosefield** [comment]
Out of curiosity, what is the stabilization process like? I know that releases happen every 6 weeks, with the next one in ETA ~20 days, but I don't know how soon things need to be merged/how long they need to sit on `main` before they can be included.

File: doc/cascadia/profiles.schema.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/cascadia/profiles.schema.json:503]: > don't really know how to confirm correctness of anything in the schema though.

lol me neither. VsCode will do it, but more often than not, I just wait for someone to complain about it being wrong 😅

**Comment by zadjii-msft** [comment]
> what is the stabilization process like

Generally we just cut a release with whatever's in `main` roughly every 6 weeks. Most of everything in Preview gets moved to the Stable branch at that time too. Usually we'll cut the build about a week before the release gets pushed to the Store. At that time we'll internally bash on it to see what breaks, and try to quick make some hotfixes before the release gets pushed. I'll admit though, it's more of an art than a strict engineering process 😄  

I'd suspect that this and your other PRs will all land in 1.11 pretty easily ☺️

File: src/cascadia/TerminalApp/AppCommandlineArgs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rosefield) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:311]: Would it be fine to just merge the command line portion and have that call the separate actions, or should there be a higher level of integration?

[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:311]: Okay so taking a look at the [man pages](https://man7.org/linux/man-pages/man1/tmux.1.html):

<details>
<summary>tmux man page selection</summary>

```
     join-pane [-bdfhv] [-l size] [-s src-pane] [-t dst-pane]
                   (alias: joinp)
             Like split-window, but instead of splitting dst-pane and
             creating a new pane, split it and move src-pane into the
             space.  This can be used to reverse break-pane.  The -b
             option causes src-pane to be joined to left of or above
             dst-pane.
     move-pane [-bdfhv] [-l size] [-s src-pane] [-t dst-pane]
                   (alias: movep)
             Does the same as join-pane.
     select-pane [-DdeLlMmRUZ] [-T title] [-t target-pane]
                   (alias: selectp)
             Make pane target-pane the active pane in its window.  If
             one of -D, -L, -R, or -U is used, respectively the pane
             below, to the left, to the right, or above the target pane
             is used...
     swap-pane [-dDUZ] [-s src-pane] [-t dst-pane]
                   (alias: swapp)
             Swap two panes.  If -U is used and no source pane is
             specified with -s, dst-pane is swapped with the previous
             pane (before it numerically); -D swaps with the next pane
             (after it numerically).
```
</details>

So they definitely don't allow `move-pane -t 4 left`, because that's nonsensical. "move the pane to tab 4, then move it left"? 

Maybe we should follow their lead on this one. 
* `swap-pane [up|down|left|right|prev|nextInOrder|prevInOrder]|[-t,--target tgt-pane]`
  * This is what we've already got as `move-pane`, so this would be a rename, but without a short form (sad, but okay)
  * obviously, we'd add the `nextInOrder` (etc) values after #10909
  * and we'd add the `-t` argument in another PR, as an exclusive arg with the direction
* `move-pane [-t,--tab index]`: move the pane to the given tab. If `index` is `new`, then moves it to a new tab.
  * in the future we'd add `[-w,--window id]` for moving to another window. 
  * we could add additional params for like, `[-d,--destination pane-id]`: split the given destination pane rather than the active pane in the tab

and I suppose with both we could ass a `[-s,--source pane-id]` arg as well.

I might just be openly brainstorming at this point. What do you think? we haven't shipped `move-pane` yet, so we're definitely still free to change it if we feel there's a better name

[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:311]: > Would it be fine to just merge the command line portion and have that call the separate actions, or should there be a higher level of integration?

Yea we definitely thought it was okay to do that. Having two actions internally for one subcommand is _kinda weird_, but is definitely easier to deal with logically than trying to combine them into one args blob with optional properties and all that.

[the remark this answers is not in the document]
Comment (Rosefield) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:311]: I'm happy to rename `move-pane` to `swap-pane` and `move-pane-to-tab` to `move-pane,` or to just make it all `move-pane` just let me know. Bit annoying to dig through all the resource files again but its not too bad.

[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp:311]: FWIW, the team did feel that we should just do it all on `move-pane`, I've just been brainstorming myself over here since re-reading this thread this morning.

I'm trying to think of all the moving things we want to do with panes, and coming up with a single command that looks like... this?
```
move-pane(mp) [--window,-w window-id] [--source,-s src-pane] [--tab,-t index]|[up|down|left|right|prev|nextInOrder|prevInOrder]|[--destination,-d dest-pane]
```

* we want to move panes to other tabs
* we want to move panes to other windows
* we want to swap a pane with another pane, by ID
* we want to move a pane in a given direction (we want to swap it in a direction)
* we want to be able to move the active pane, or a specific pane in a tab

and I think all that together means that we want to 
1. swap with a (source)? and (a tgt or a direction)
2. move with a (source)? and (a tab)? and (a tgt pane)?

and if these were all on one command, then I don't think it works.  We'd need `tab` and `direction` to be exclusive to differentiate between 1 and 2. But then `direction` and `dest` would need to be exclusive for swap, but not exclusive for move, and now the rules don't work.

So in conclusion, yea:

> rename `move-pane` to `swap-pane` and `move-pane-to-tab` to `move-pane`

(sorry for the runaround on all this!)

**Comment by Rosefield** [comment]
Renamed `move-pane` -> `swap-pane` and `move-pane-to-tab` -> `move-pane`

**Review by zadjii-msft — APPROVED** [review]
Thank you for putting up with our very organic planning process on this one ☺️

**Comment by Rosefield** [comment]
I think the labels on this can be updated now? [Needs-Second] and remove [Needs-Discussion]?

**Comment by zadjii-msft** [comment]
Yep, my bad 😋

**Review by carlos-zamora — APPROVED** [review]
This looks great! Thanks!

There's a few `const` that can be sprinkled around, but that's all I found.

**Comment by carlos-zamora** [comment]
Oh! Don't forget to update the docs too! https://github.com/MicrosoftDocs/terminal

**Comment by Rosefield** [comment]
Created documentation updates for all of the features I worked on.

**Comment by Rosefield** [comment]
I believe this is ready to merge?

**Comment by carlos-zamora** [comment]
@msftbot merge this in 5 minutes

**Review by DHowett — COMMENTED** [review]
Just wanted to say that this is really, really clever. I love it. Well done and thank you!
